gabi_test.go

```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+	rhs.Mul(new(big.Int).Exp(totalP, challenge, testPubK.N), totalComm).Mod(rhs, testPubK.N)
+	require.Equal(t, 0, rhs.Cmp(lhs))
+
+	// Build and verify the proofs
+	proofPs := make([]*ProofP, len(builders))
+	var kss []string // specifies to Verify() below which proof uses which keyshare server
+	for i := range builders {
+		if keyNames[i] == nil {
+			kss = append(kss, "")
+		} else {
+			proofPs[i] = proofP
+			kss = append(kss, "keyshare server")
+		}
+	}
+	proofs, err := builders.BuildDistributedProofList(challenge, proofPs)
+	require.NoError(t, err)
+	require.True(t, proofs.Verify(keysSlice, context, nonce, signature, kss))
+
+	// In case of issuance, check that we can construct valid credentials
+	privkeys := map[string]*gabikeys.PrivateKey{
+		"testPubK": testPrivK, "testPubK1": testPrivK1, "testPubK2": testPrivK2,
+	}
+	for i, proof := range proofs {
+		if proofU, ok := proof.(*ProofU); ok {
+			builder := builders[i].(*CredentialBuilder)
+			issuer := NewIssuer(privkeys[builder.pk.Issuer], builder.pk, context)
+			attrs := slices.Clone(testAttributes1)
+			var randomblind []int
+			for i := range proofU.MUserResponses {
+				randomblind = append(randomblind, i-1)
+				attrs[i-1] = nil
+			}
+			ism, err := issuer.IssueSignature(proofU.U, attrs, nil, builder.nonce2, randomblind)
+			require.NoError(t, err, "error creating Issue Signature")
+			cred, err := builder.ConstructCredential(ism, attrs)
+			require.NoError(t, err, "error creating credential")
+			cred.Signature.Verify(builder.pk, cred.Attributes)
+		}
+	}
+}
+
+var context = bigOne
+
+func newCredbuilder(t *testing.T, pk *gabikeys.PublicKey, secret, ourP *big.Int, randomblind []int) *CredentialBuilder {
+	nonce2, err := common.RandomBigInt(gabikeys.DefaultSystemParameters[1024].Lstatzk)
+	require.NoError(t, err)
+	credBuilder, err := NewCredentialBuilder(pk, context, secret, nonce2, ourP, randomblind)
+	require.NoError(t, err)
+	return credBuilder
+}
+
+func newDisclosureBuilder(t *testing.T,
+	sk *gabikeys.PrivateKey,
+	pk *gabikeys.PublicKey,
+	secret *big.Int,
+	ourP *big.Int,
+	rev bool,
+	rangeStatements map[int][]*rangeproof.Statement,
+) *DisclosureProofBuilder {
+	var witness *revocation.Witness
+	attrs := testAttributes1
+	if rev {
+		witness, _, _ = setupRevocation(t, sk, pk)
+		attrs = append(testAttributes1, witness.E)
+	}
+
+	cred := createKeyshareCredential(t, context, secret, ourP, attrs, NewIssuer(sk, pk, context))
+	cred.NonRevocationWitness = witness
+	require.True(t, cred.Signature.Verify(pk, cred.Attributes))
+
+	if rev {
+		require.NoError(t, cred.NonrevPrepareCache())
+		require.NoError(t, cred.NonRevocationWitness.Verify(pk))
+	}
+
+	disclosureBuilder, err := cred.CreateDisclosureProofBuilder([]int{1}, rangeStatements, rev)
+	require.NoError(t, err)
+	return disclosureBuilder
+}
+
+func ourP(ourSecret *big.Int, pk *gabikeys.PublicKey) *big.Int {
+	return new(big.Int).Exp(pk.R[0], ourSecret, pk.N)
+}
+
+func TestKeyshareResponseSingleBase(t *testing.T) {
```

Comment:
What's the added value of this test when there's already the other flexible one? I compared the logic but cannot see what's tested more here what could not already be tested in the other one.

Reply:
Yes, everything that this test covers is also covered in the other more flexible one. But the other one always uses the full machinery of disclosure and/or credential builders, while this one purely tests the new keysharing for just a single secret and base. So if this test would fail, then you can be sure that the problem is somewhere in the new keyshare code, while if the other one fails, then you can't (without further investigation) be sure if the error is there, or in the code where the keysharing interacts with proof builders.

If you feel that that benefit is too little then I would be fine with removing this test entirely.